### PR TITLE
Optional custom scheduler for KafkaMetrics

### LINKED ISSUE #4976: KafkaClientMetrics should provide possibility to use custom scheduler for binding metrics
**Please describe the feature request.**
Currently, [`KafkaClientMetrics`](https://github.com/micrometer-metrics/micrometer/blob/main/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaClientMetrics.java) does not provide the possibility to use a custom scheduler for binding metrics. Instead, it uses a default one:
```
ScheduledExecutorService scheduler = Executors
    .newSingleThreadScheduledExecutor(new NamedThreadFactory("micrometer-kafka-metrics"));
```
 Ideally, it would be great to have the ability to pass an optional custom thread pool, so later Spring-Kafka will be able to propagate optional scheduler into `KafkaClientMetrics`.

**Rationale**
I have a Spring-Boot application with multiple consumers with the same consumer group using `@KafkaListener` from Spring-Kafka. After investigation, I discovered that per each consumer it creates additional Micrometer metrics daemon thread with the name `micrometer-kafka-metrics`, which is created by `io.micrometer:micrometer.core` inside [`KafkaMetrics`](https://github.com/micrometer-metrics/micrometer/blob/main/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaMetrics.java).

The chain of invocations is the following. `KafkaMetricsAutoConfiguration` (from `spring-boot-actuator-autoconfigure`) creates a single instance of `MicrometerConsumerListener` (from Spring-Kafka) per application, and after that `MicrometerConsumerListener` has method `consumerAdded(..)` that is invoked per each Consumer. It creates `KafkaClientMetrics` (extended from `KafkaMetrics`). As a result, the number of consumers is the same as the number of threads `micrometer-kafka-metrics`.

If we have consumer factory concurrency as `N` (`new ConcurrentKafkaListenerContainerFactory<>().setConcurrency(N);`) and number of `@KafkaListener` as `M` (one per each topic, as we have `M` topics), as a result we have `N * M` threads with name `micrometer-kafka-metrics`. In most cases `N * M` is not a big number, but in some rare cases, it might be a few hundred (like in my case). I want to leave the number of concurrent consumers as is, but decrease the number of metrics threads, cause metrics threads are quite lightweight and scheduled with a fixed rate (1 minute by default).

So I'm looking at ways how to decrease the number of these metrics threads. Ideally, it would be great to have the ability to pass there optional custom thread pool, that will be shareable for all consumers and with the desired number of threads. 

**Additional context**
I created a demo application to show this issue with the management API endpoint that returns current threads. You could execute the following [integration test](https://github.com/vasiliy-sarzhynskyi/spring-kafka-micrometer-metrics-threads-demo/blob/main/src/test/java/com/vsarzhynskyi/spring/kafka/micrometer/metrics/threads/demo/ThreadManagementControllerIntegrationTest.java#L66-L83) by your own to see the number of created threads `micrometer-kafka-metrics`.

**References**
- [Clarification on StackOverflow](https://stackoverflow.com/questions/78334547/kafka-metrics-for-spring-kafka-consumers-possibility-to-decrease-number-of-thr)
- [Demo](https://github.com/vasiliy-sarzhynskyi/spring-kafka-micrometer-metrics-threads-demo)
- [Spring-Kafka creates `KafkaClientMetrics`](https://github.com/spring-projects/spring-kafka/blob/main/spring-kafka/src/main/java/org/springframework/kafka/core/MicrometerConsumerListener.java)

### PR DESCRIPTION
**Pull request by vasiliy-sarzhynskyi**
Resolves gh-4976

### CODE REVIEW

**Comment by vasiliy-sarzhynskyi** [comment]
It's a proposed contribution for [enhancement #4976](https://github.com/micrometer-metrics/micrometer/issues/4976) to extend `KafkaMetrics` with custom scheduler

**Review by shakuzen — COMMENTED** [review]
Thank you for the pull request. This looks reasonable to me. Before merging, I'm trying to get an idea of the impact to other projects so this can be most easily used by end users. What would change in Spring Kafka and Spring Boot? I guess `MicrometerConsumerListener` in Spring Kafka would need to optionally take a `ScheduledExecutorService` and `KafkaMetricsAutoConfiguration` in Spring Boot would need a way to get a `ScheduledExecutorService` from user configuration (how would it do this?) to pass to the `MicrometerConsumerListener` it makes.

@artembilan do you have any concerns about this from a Spring Kafka perspective?

**Review by artembilan — CHANGES_REQUESTED** [review]

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaClientMetrics.java [hunk 1]
```diff
@@ -60,6 +62,16 @@ public KafkaClientMetrics(Producer<?, ?> kafkaProducer) {
         super(kafkaProducer::metrics);
     }
 
+    /**
+     * Kafka {@link Consumer} metrics binder
+     * @param kafkaConsumer consumer instance to be instrumented
+     * @param tags additional tags
+     * @param scheduler scheduler to check and bind metrics
+     */
+    public KafkaClientMetrics(Consumer<?, ?> kafkaConsumer, Iterable<Tag> tags, ScheduledExecutorService scheduler) {
```
[the remark this answers is not in the document]
Comment (shakuzen) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaClientMetrics.java:71]: Thanks for the review @artembilan.
@vasiliy-sarzhynskyi any reason to not add similar constructors for producers and admin clients?

[the remark this answers is not in the document]
Comment (vasiliy-sarzhynskyi) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaClientMetrics.java:71]: @shakuzen I need to check how producers and admin clients are used by Spring-Kafka, and how many threads we might have with multiple produced topics (whether the number of threads increased or not).  We could add there as well, but do we have use cases when producers and admin clients might need that? In my understanding, it's especially relevant for consumers as we might specify `concurrency` config value, and a single app instance will have hundreds of threads.

[the remark this answers is not in the document]
Comment (artembilan) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaClientMetrics.java:71]: I’m not sure how is this relevant to the use of those clients in the app, the scheduler we provide here is used for publishing metrics from those clients. See `bindTo()` impl.

[the remark this answers is not in the document]
Comment (shakuzen) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaClientMetrics.java:71]: This scheduler isn't for publishing; MeterRegistry implementations can be configured with a scheduler to publish metrics. This scheduler is for syncing the metrics from the Kafka client to Micrometer's registry periodically. The background is in https://github.com/micrometer-metrics/micrometer/issues/4976, but with many clients and the existing code, each client will have a thread for syncing between the Kafka client's metrics and Micrometer. With the changes here, a thread pool could be shared among the clients for this syncing.

[the remark this answers is not in the document]
Comment (vasiliy-sarzhynskyi) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaClientMetrics.java:71]: Thank you. In that case, I will update the current PR for `Producer` and `Admin` as well

[the remark this answers is not in the document]
Comment (vasiliy-sarzhynskyi) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaClientMetrics.java:81]: @shakuzen I pushed my changes for `Producer` and `AdminClient` constructors as well. After I rebased on branch `main`, recently added test class `VirtualThreadMetricsTests` is failing, looks like it's flaky.
I could try to increase the upper bound threshold inside `VirtualThreadMetricsTests pinnedEventsShouldBeRecorded(..)`, but would be better to contact owner of these changes to clarify the intention of asserted values

[the remark this answers is not in the document]
Comment (shakuzen) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaClientMetrics.java:81]: Pinging @jonatan-ivanov about the flakiness of `VirtualThreadMetricsTests`. I'll re-run the CI manually in the meantime.

**Review by shakuzen — APPROVED** [review]
Thanks for the pull request and quick response to review. This looks good to me. I'll polish a couple things when merging.

**Comment by shakuzen** [comment]
I've added the same changes for `KafkaStreamsMetrics` and updated the JavaDocs to specifically mention the purpose and responsibilities for the custom scheduler.

**Pull request merged**